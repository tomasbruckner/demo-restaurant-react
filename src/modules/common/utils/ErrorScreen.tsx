import React, { PureComponent } from 'react';
import PageLayout from '../layouts/PageLayout';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';

export class ErrorScreen extends PureComponent<{ title: string } & WithTranslation> {
  render() {
    const { t } = this.props;

    return <PageLayout title={t('page.error.title')} />;
  }
}

export default withTranslation(I18N_CORE)(ErrorScreen);
