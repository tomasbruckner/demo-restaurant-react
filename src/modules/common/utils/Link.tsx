import { Link as RouterLink } from 'react-router-dom';

export default RouterLink;
