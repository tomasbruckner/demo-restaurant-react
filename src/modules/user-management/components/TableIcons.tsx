import React, { forwardRef } from 'react';
import AddBox from '@material-ui/icons/AddBox';
import Check from '@material-ui/icons/Check';
import Clear from '@material-ui/icons/Clear';
import DeleteOutline from '@material-ui/icons/DeleteOutline';
import ChevronRight from '@material-ui/icons/ChevronRight';
import Edit from '@material-ui/icons/Edit';
import SaveAlt from '@material-ui/icons/SaveAlt';
import FilterList from '@material-ui/icons/FilterList';
import FirstPage from '@material-ui/icons/FirstPage';
import LastPage from '@material-ui/icons/LastPage';
import ChevronLeft from '@material-ui/icons/ChevronLeft';
import Search from '@material-ui/icons/Search';
import ArrowDownward from '@material-ui/icons/ArrowDownward';
import Remove from '@material-ui/icons/Remove';
import ViewColumn from '@material-ui/icons/ViewColumn';

// eslint-disable-next-line import/prefer-default-export
export const tableIcons = {
  // @ts-ignore
  Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} data-cy="table-add" />),
  // @ts-ignore
  Check: forwardRef((props, ref) => <Check {...props} ref={ref} data-cy="table-check" />),
  // @ts-ignore
  Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} data-cy="table-clear" />),
  // @ts-ignore
  Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} data-cy="table-delete" />),
  DetailPanel: forwardRef((props, ref) => (
    // @ts-ignore
    <ChevronRight {...props} ref={ref} data-cy="table-detail-panel" />
  )),
  // @ts-ignore
  Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} data-cy="table-edit" />),
  // @ts-ignore
  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} data-cy="table-export" />),
  // @ts-ignore
  Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} data-cy="table-filter" />),
  FirstPage: forwardRef((props, ref) => (
    // @ts-ignore
    <FirstPage {...props} ref={ref} data-cy="table-first-page" />
  )),
  // @ts-ignore
  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} data-cy="table-last-page" />),
  NextPage: forwardRef((props, ref) => (
    // @ts-ignore
    <ChevronRight {...props} ref={ref} data-cy="table-next-page" />
  )),
  PreviousPage: forwardRef((props, ref) => (
    // @ts-ignore
    <ChevronLeft {...props} ref={ref} data-cy="table-previous-page" />
  )),
  ResetSearch: forwardRef((props, ref) => (
    // @ts-ignore
    <Clear {...props} ref={ref} data-cy="table-reset-search" />
  )),
  // @ts-ignore
  Search: forwardRef((props, ref) => <Search {...props} ref={ref} data-cy="table-search" />),
  SortArrow: forwardRef((props, ref) => (
    // @ts-ignore
    <ArrowDownward {...props} ref={ref} data-cy="table-sort-arrow" />
  )),
  ThirdStateCheck: forwardRef((props, ref) => (
    // @ts-ignore
    <Remove {...props} ref={ref} data-cy="table-third-state-check" />
  )),
  ViewColumn: forwardRef((props, ref) => (
    // @ts-ignore
    <ViewColumn {...props} ref={ref} data-cy="table-view-column" />
  )),
};
