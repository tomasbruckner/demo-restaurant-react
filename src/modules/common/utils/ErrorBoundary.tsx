import React, { PureComponent } from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import Logger from 'utils/logger';
import { I18N_CORE } from 'core/i18n/i18n';
import ErrorScreen from './ErrorScreen';

export class ErrorBoundaryComp extends PureComponent<
  { renderProp?: () => React.ReactNode } & WithTranslation
> {
  state = {
    error: null,
  };

  static getDerivedStateFromError(error: Error) {
    return { error };
  }

  componentDidCatch(error: unknown, errorInfo: unknown) {
    Logger.error(error, errorInfo);
  }

  render() {
    const { children, renderProp, t } = this.props;
    const { error } = this.state;

    if (!error) {
      return children;
    }

    if (renderProp) {
      return renderProp();
    }

    return <ErrorScreen title={t('CommonComponents.ErrorBoundary.error')} />;
  }
}

export default withTranslation(I18N_CORE)(ErrorBoundaryComp);
