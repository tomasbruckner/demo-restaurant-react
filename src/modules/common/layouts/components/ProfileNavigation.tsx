import React, { PureComponent } from 'react';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Link from 'modules/common/utils/Link';
import IconButton from '@material-ui/core/IconButton';
import AccountCircle from '@material-ui/icons/AccountCircle';
import { LOGIN, USER_PROFILE } from 'core/router/urls';
import { RouteComponentProps, withRouter } from 'react-router';
import { logOut } from 'core/redux/ducks/UserDuck';
import { connect, ConnectedProps } from 'react-redux';

const connector = connect(null, {
  dispatchLogOut: logOut,
});

class ProfileNavigationComponent extends PureComponent<
  WithTranslation & ConnectedProps<typeof connector> & RouteComponentProps
> {
  state = {
    anchorEl: null,
  };

  onMenuClose = () => {
    this.setState({ anchorEl: null });
  };

  onProfileClick = ({ currentTarget }: React.SyntheticEvent) => {
    this.setState({ anchorEl: currentTarget });
  };

  onLogoutClick = () => {
    const { history, dispatchLogOut } = this.props;

    dispatchLogOut();
    history.push(LOGIN);
  };

  render() {
    const { anchorEl } = this.state;
    const { t } = this.props;

    return (
      <>
        <IconButton
          edge="end"
          aria-label="account of current user"
          aria-haspopup="true"
          color="inherit"
          onClick={this.onProfileClick}
        >
          <AccountCircle />
        </IconButton>
        <Menu
          anchorEl={anchorEl}
          anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
          keepMounted
          transformOrigin={{ vertical: 'top', horizontal: 'right' }}
          open={!!anchorEl}
          onClose={this.onMenuClose}
        >
          <MenuItem to={USER_PROFILE} component={Link}>
            {t('component.navigation.menu.profile')}
          </MenuItem>
          <MenuItem onClick={this.onLogoutClick}>{t('component.navigation.menu.logout')}</MenuItem>
        </Menu>
      </>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(withRouter(ProfileNavigationComponent)));
