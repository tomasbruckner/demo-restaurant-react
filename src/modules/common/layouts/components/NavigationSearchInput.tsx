import React, { PureComponent } from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { RootState } from 'core/redux/store';
import {
  getSearchSelector,
  searchResultFilterAsync,
  clearSearch,
} from 'core/redux/ducks/SearchDuck';
import SearchIcon from '@material-ui/icons/Search';
import { fade, withStyles } from '@material-ui/core/styles';
import styled from 'styled-components/macro';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Autocomplete from '@material-ui/lab/Autocomplete';
import CircularProgress from '@material-ui/core/CircularProgress';
import Input from 'modules/common/components/controls/Input';
import Typography from 'modules/common/components/controls/Typography';
import { RouteComponentProps, withRouter } from 'react-router';
import { getRestaurantDetail } from 'core/router/urls';
import { DEFAULT_SEARCH_PAGE_SIZE } from 'utils/constants';

const SearchIconWrapper = styled.div`
  padding-left: 16px;
  padding-right: 16px;
  height: 100%;
  position: absolute;
  pointer-events: none;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const SearchWrapper = styled.div`
  position: relative;
  border-radius: 3px;
  margin-left: 32px;
  background-color: ${p => {
    return `${fade(p.theme.white, 0.15)}`;
  }};
  '&:hover': {
    background-color: ${p => `${fade(p.theme.white, 0.25)}`};
  },
  margin-left: 0;
  width: 300px;
`;

const StyledInput = withStyles({
  root: {
    color: 'white',
    '&:focus': {
      color: 'white',
    },
  },
})(Input);

const connector = connect(
  (state: RootState) => {
    return {
      searchResult: getSearchSelector(state),
    };
  },
  {
    dispatchSearch: searchResultFilterAsync,
    dispatchClearSearch: clearSearch,
  },
);

class NavigationSearchInput extends PureComponent<
  WithTranslation & ConnectedProps<typeof connector> & RouteComponentProps
> {
  state = {
    isLoading: false,
    isOpen: false,
    searchValue: '',
  };

  public componentDidMount() {
    const { dispatchClearSearch } = this.props;

    dispatchClearSearch();
  }

  onToggleOpen = () => {
    const { isOpen } = this.state;

    this.setState({ isOpen: !isOpen, searchValue: '' });
  };

  redirect = (restaurantId: number) => {
    const { history } = this.props;

    history.push(getRestaurantDetail(restaurantId));
  };

  onChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
    const { dispatchSearch } = this.props;

    this.setState({ isLoading: true, searchValue: event.target.value });
    try {
      await dispatchSearch({
        take: DEFAULT_SEARCH_PAGE_SIZE,
        skip: 0,
        sort: {
          predicate: 'name',
          reverse: false,
        },
        filter: {
          name: event.target.value,
        },
      });
    } finally {
      this.setState({ isLoading: false });
    }
  };

  render() {
    const { searchResult } = this.props;
    const { isLoading, isOpen, searchValue } = this.state;

    return (
      <SearchWrapper>
        {!isOpen && (
          <SearchIconWrapper>
            <SearchIcon />
          </SearchIconWrapper>
        )}
        <Autocomplete
          style={{ width: 300 }}
          open={isOpen}
          onOpen={this.onToggleOpen}
          onClose={this.onToggleOpen}
          options={searchResult.data}
          loading={isLoading}
          getOptionLabel={o => o.name}
          clearOnBlur
          clearOnEscape
          inputValue={searchValue}
          renderOption={option => (
            <Typography onClick={() => this.redirect(option.restaurantId)}>
              {option.name}
            </Typography>
          )}
          renderInput={params => (
            <StyledInput
              {...params}
              fullWidth
              variant="outlined"
              onChange={this.onChange}
              InputProps={{
                ...params.InputProps,
                style: {
                  color: 'white',
                },
                endAdornment: (
                  <>
                    {isLoading ? <CircularProgress color="inherit" size={20} /> : null}
                    {params.InputProps.endAdornment}
                  </>
                ),
              }}
            />
          )}
        />
      </SearchWrapper>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(withRouter(NavigationSearchInput)));
