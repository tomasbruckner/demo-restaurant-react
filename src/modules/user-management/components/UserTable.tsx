import React, { PureComponent } from 'react';
import MaterialTable from 'material-table';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import { Role } from 'utils/constants';
import { tableIcons } from './TableIcons';
import { connect, ConnectedProps } from 'react-redux';
import { RootState } from 'core/redux/store';
import { userListSelector, userFilterAsync } from 'core/redux/ducks/UserDuck';
import UserService from 'services/userService';

const connector = connect(
  (state: RootState) => {
    return {
      userList: userListSelector(state),
    };
  },
  {
    fetchUsers: userFilterAsync,
  },
);

export class UserTable extends PureComponent<WithTranslation & ConnectedProps<typeof connector>> {
  render() {
    const { t } = this.props;

    return (
      <>
        <MaterialTable
          // @ts-ignore
          icons={tableIcons}
          columns={[
            {
              title: '',
              field: 'userId',
              hidden: true,
            },
            {
              title: t('page.user-management.table.columns.last-name'),
              field: 'lastName',
            },
            {
              title: t('page.user-management.table.columns.first-name'),
              field: 'firstName',
            },
            {
              title: t('page.user-management.table.columns.email'),
              field: 'email',
              editable: 'never',
            },
            {
              title: t('page.user-management.table.columns.role'),
              field: 'roleId',
              lookup: {
                [Role.User]: t('enum.role.user'),
                [Role.Owner]: t('enum.role.owner'),
                [Role.Admin]: t('enum.role.admin'),
              },
            },
          ]}
          data={async query => {
            const filters = query.filters.reduce(
              (acc, x) => {
                const key = x.column.field as 'email' | 'lastName' | 'firstName' | 'roleId';
                acc[key] = x.value;

                return acc;
              },
              {
                email: undefined,
                lastName: undefined,
                firstName: undefined,
                roleId: undefined,
              },
            );
            const response = await UserService.filter({
              sort: {
                predicate: query.orderBy?.field?.toString() || 'lastName',
                reverse: query.orderDirection === 'desc',
              },
              take: query.pageSize,
              skip: query.page * query.pageSize,
              filter: {
                email: filters.email,
                firstName: filters.firstName,
                lastName: filters.lastName,
                roleIdList: filters.roleId,
              },
            });
            return {
              data: response.data,
              page: query.page,
              totalCount: response.size,
            };
          }}
          options={{
            search: false,
            showTitle: false,
            draggable: false,
            filtering: true,
            pageSize: 10,
          }}
          editable={{
            onRowUpdate: async ({ userId, roleId, firstName, lastName }) => {
              await UserService.editUser(userId, {
                roleId,
                firstName,
                lastName,
              });
            },
            onRowDelete: async ({ userId }) => {
              await UserService.deleteUser(userId);
            },
          }}
        />
      </>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(UserTable));
