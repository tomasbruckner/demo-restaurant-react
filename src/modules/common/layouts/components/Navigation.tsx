import React, { PureComponent } from 'react';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from 'modules/common/components/controls/Typography';
import Link from 'modules/common/utils/Link';
import styled from 'styled-components/macro';
import { WithTranslation, withTranslation } from 'react-i18next';
import { I18N_CORE } from 'core/i18n/i18n';
import { OWNED_RESTAURANTS, RESTAURANTS, USER_MANAGEMENT } from 'core/router/urls';
import ProfileNavigation from './ProfileNavigation';
import NavigationSearchInput from './NavigationSearchInput';
import RoleAccessRenderer from '../../utils/RoleAccessRenderer';

const FlexGrow = styled.div`
  flex-grow: 1;
`;

const LinkWrapper = styled.div<Props>`
  display: flex;
  margin-left: 32px;
`;

const StyledLink = styled(Link)`
  color: white;
  margin-right: 16px;
  text-decoration: none;
  &:visited {
    color: white;
  }

  &:active {
    color: white;
  }

  &:hover {
    color: white;
  }

  &:link {
    color: white;
  }
`;

type Props = {};

class NavigationComponent extends PureComponent<Props & WithTranslation> {
  render() {
    const { t } = this.props;

    return (
      <FlexGrow>
        <AppBar position="static">
          <Toolbar>
            <Typography variant="h6" noWrap>
              {t('component.navigation.home')}
            </Typography>
            <NavigationSearchInput />
            <div />
            <LinkWrapper>
              <Typography component={StyledLink} to={RESTAURANTS}>
                {t('component.navigation.links.restaurants')}
              </Typography>
              <RoleAccessRenderer adminOrOwner>
                <Typography component={StyledLink} to={OWNED_RESTAURANTS}>
                  {t('component.navigation.links.owned-restaurants')}
                </Typography>
              </RoleAccessRenderer>
              <RoleAccessRenderer onlyAdmin>
                <Typography component={StyledLink} to={USER_MANAGEMENT}>
                  {t('component.navigation.links.users')}
                </Typography>
              </RoleAccessRenderer>
            </LinkWrapper>
            <FlexGrow />
            <div>
              <ProfileNavigation />
            </div>
          </Toolbar>
        </AppBar>
      </FlexGrow>
    );
  }
}

export default withTranslation(I18N_CORE)(NavigationComponent);
