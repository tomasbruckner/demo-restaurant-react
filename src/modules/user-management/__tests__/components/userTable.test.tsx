import React from 'react';
import 'jest-styled-components';
import { UserTable } from 'modules/user-management/components/UserTable';
import { fakeFunction, fakeTranslationProps } from 'utils/testUtils';
import { shallowToJson } from 'enzyme-to-json';
import { shallow } from 'enzyme';

describe('UserTable', () => {
  it('basic example', () => {
    const tree = shallowToJson(
      shallow(<UserTable fetchUsers={fakeFunction} userList={[]} {...fakeTranslationProps} />),
    );

    expect(tree).toMatchSnapshot();
  });
});
