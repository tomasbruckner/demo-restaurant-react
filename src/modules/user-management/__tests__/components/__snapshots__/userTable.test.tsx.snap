// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable basic example 1`] = `
<Fragment>
  <WithStyles(Component)
    columns={
      Array [
        Object {
          "field": "userId",
          "hidden": true,
          "title": "",
        },
        Object {
          "field": "lastName",
          "title": "page.user-management.table.columns.last-name",
        },
        Object {
          "field": "firstName",
          "title": "page.user-management.table.columns.first-name",
        },
        Object {
          "editable": "never",
          "field": "email",
          "title": "page.user-management.table.columns.email",
        },
        Object {
          "field": "roleId",
          "lookup": Object {
            "1": "enum.role.admin",
            "2": "enum.role.owner",
            "3": "enum.role.user",
          },
          "title": "page.user-management.table.columns.role",
        },
      ]
    }
    data={[Function]}
    editable={
      Object {
        "onRowDelete": [Function],
        "onRowUpdate": [Function],
      }
    }
    icons={
      Object {
        "Add": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Check": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Clear": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Delete": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "DetailPanel": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Edit": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Export": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Filter": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "FirstPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "LastPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "NextPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "PreviousPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "ResetSearch": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Search": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "SortArrow": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "ThirdStateCheck": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "ViewColumn": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      }
    }
    options={
      Object {
        "draggable": false,
        "filtering": true,
        "pageSize": 10,
        "search": false,
        "showTitle": false,
      }
    }
  />
</Fragment>
`;
